File: js/primeGame.js
function isWinner(x, nums) {
  if (!nums || x < 1) return null;
  const maxN = Math.max(...nums);
  const sieve = new Array(maxN + 1).fill(true);
  sieve[0] = sieve[1] = false;

  for (let i = 2; i * i <= maxN; i++) {
    if (sieve[i]) {
      for (let j = i * i; j <= maxN; j += i) {
        sieve[j] = false;
      }
    }
  }

  const primes = [];
  for (let i = 2; i <= maxN; i++) {
    if (sieve[i]) primes.push(i);
  }

  let mariaWins = 0;
  let benWins = 0;

  for (let i = 0; i < x; i++) {
    const n = nums[i];
    let primeCount = 0;

    for (const prime of primes) {
      if (prime > n) break;
      primeCount++;
    }

    if (primeCount % 2 === 1) {
      mariaWins++;
    } else {
      benWins++;
    }
  }

  if (mariaWins > benWins) {
    return "Maria";
  } else if (benWins > mariaWins) {
    return "Ben";
  } else {
    return null;
  }
}

const x = 3;
const nums = [4, 5, 1];
console.log(isWinner(x, nums));

File: python/primeGame.py
def isWinner(x, nums):
    if not nums or x < 1:
        return None
    max_n = max(nums)
    sieve = [True] * (max_n + 1)
    sieve[0] = sieve[1] = False
    for i in range(2, int(max_n ** 0.5) + 1):
        if sieve[i]:
            for j in range(i * i, max_n + 1, i):
                sieve[j] = False

    primes = [i for i, is_prime in enumerate(sieve) if is_prime]
    maria_wins = 0
    ben_wins = 0

    for n in nums:
        prime_count = 0
        for prime in primes:
            if prime > n:
                break
            prime_count += 1
        if prime_count % 2 == 1:
            maria_wins += 1
        else:
            ben_wins += 1
    if maria_wins > ben_wins:
        return "Maria"
    elif ben_wins > maria_wins:
        return "Ben"
    else:
        return None
